### PR TITLE
[ReplayBuffer] - Classifier - Regressor

### PR DESCRIPTION
**Pull request by raphaelsty**
Hey,

ReplayBuffer stores a dictionary of tuples (characteristics, target). Tuples stored in the buffer are observations that are difficult for a model to regress or classify.

The purpose of the buffer is to re-train a model on targets that are difficult to extrapolate. This difficulty is determined using a loss function provided during model instantiation.

During fitting, the model is re-trained with probability p on one of the observations (uniformly sampled) in the buffer. The model is trained with a probability (1 - p) on the new data. This probability is constant.

Validation:

| Dataset - Metric 	| CreditCard - ROCAUC 	| Phishing - Accuracy 	| TrumpApproval - MAE 	| ToulouseBikes - MAE 	|
|-----------------------	|---------------------	|---------------------	|---------------------	|---------------------	|
| ReplayBuffer 	| **0.914474** 	| **89.68** 	| **1.847611** 	| 1.424836 	|
| Without Replay Buffer 	| 0.891071 	| 89.28 	| 2.077663 	| **0.92614** 	|
|  	|  	|  	|  	|  	|

Experience with the imbalanced dataset CreditCard:

```python
from creme import compose 
from creme import feature_extraction
from creme import stats
from creme import datasets
from creme import preprocessing
from creme import replay
from creme import linear_model
from creme import optim
from creme import metrics
from creme import model_selection

X_y = datasets.CreditCard()

model = (
    preprocessing.StandardScaler() |
    replay.ReplayBufferClassifier(
        classifier = linear_model.LogisticRegression(), 
        loss_function = optim.losses.CrossEntropy(), # Loss function to eval predictions and construct the buffer.
        p    = 0.2, # Probability to fit on an observation (x, y) of the buffer.
        size = 30, # Size of the buffer, smaller seems better.
        seed = 42,
    )
)

model_selection.progressive_val_score(
    X_y,
    model,
    metrics.ROCAUC(),
    print_every=15_000,
    show_time=True
)
```

```
# Score with the buffer:
[15,000] ROCAUC: 0.915715 – 0:00:03
[30,000] ROCAUC: 0.907942 – 0:00:08
[45,000] ROCAUC: 0.931235 – 0:00:11
[60,000] ROCAUC: 0.927454 – 0:00:15
[75,000] ROCAUC: 0.922019 – 0:00:19
[90,000] ROCAUC: 0.927155 – 0:00:22
[105,000] ROCAUC: 0.921646 – 0:00:25
[120,000] ROCAUC: 0.91881 – 0:00:29
[135,000] ROCAUC: 0.918308 – 0:00:32
[150,000] ROCAUC: 0.908016 – 0:00:36
[165,000] ROCAUC: 0.915419 – 0:00:39
[180,000] ROCAUC: 0.91489 – 0:00:43
[195,000] ROCAUC: 0.910758 – 0:00:48
[210,000] ROCAUC: 0.909915 – 0:00:52
[225,000] ROCAUC: 0.909362 – 0:00:56
[240,000] ROCAUC: 0.909802 – 0:01:00
[255,000] ROCAUC: 0.914763 – 0:01:03
[270,000] ROCAUC: 0.916702 – 0:01:07
ROCAUC: 0.914474
```

In this example, the minority class is over-represented in the buffer because it is a class that generates a large loss when trying to predict. 

Here we just train a logistic regression without the buffer:

```python
from creme import linear_model
from creme import metrics
from creme import model_selection
from creme import preprocessing
from creme import datasets


X_y = datasets.CreditCard()

model = (
    preprocessing.StandardScaler() |
    linear_model.LogisticRegression()
)

metric = metrics.ROCAUC()

model_selection.progressive_val_score(
    X_y,
    model,
    metric,
    print_every=15_000,
    show_time=True
)

```

```
# Score without the buffer:
[15,000] ROCAUC: 0.899341 – 0:00:03
[30,000] ROCAUC: 0.87079 – 0:00:07
[45,000] ROCAUC: 0.899804 – 0:00:12
[60,000] ROCAUC: 0.89192 – 0:00:16
[75,000] ROCAUC: 0.890126 – 0:00:20
[90,000] ROCAUC: 0.897645 – 0:00:24
[105,000] ROCAUC: 0.889682 – 0:00:28
[120,000] ROCAUC: 0.886271 – 0:00:32
[135,000] ROCAUC: 0.883233 – 0:00:35
[150,000] ROCAUC: 0.885329 – 0:00:38
[165,000] ROCAUC: 0.897751 – 0:00:41
[180,000] ROCAUC: 0.896706 – 0:00:45
[195,000] ROCAUC: 0.896068 – 0:00:49
[210,000] ROCAUC: 0.894425 – 0:00:51
[225,000] ROCAUC: 0.893745 – 0:00:55
[240,000] ROCAUC: 0.893375 – 0:00:57
[255,000] ROCAUC: 0.89189 – 0:01:00
[270,000] ROCAUC: 0.893778 – 0:01:03
ROCAUC: 0.891071
```

By adjusting weights several times on observations associated with higher loss, the BufferReplayClassifier manages to improve the AUC compared to the LogisticRegression alone. 

I created a BufferReplayRegressor and a Classifier. This is a draft PR, which means the work is in progress because I need to add unit tests. I think these classes deserve more testing on different datasets.

#### datasets.Phishing:

With ReplayBuffer:

```python
from creme import compose 
from creme import feature_extraction
from creme import stats
from creme import datasets
from creme import preprocessing
from creme import replay
from creme import linear_model
from creme import optim
from creme import metrics
from creme import model_selection

model = (
    preprocessing.StandardScaler() |
    replay.ReplayBufferClassifier(
        classifier = linear_model.LogisticRegression(), 
        loss_function = optim.losses.CrossEntropy(), 
        p = 0.1, # This configuration brings a slight improvement.
        size = 40,  # Smaller is better.
        seed = 42,
    )
    
)

X_y = datasets.Phishing()

model_selection.progressive_val_score(
    X_y,
    model,
    metrics.Accuracy(),
    print_every=200,
    show_time=True
)
```

```
# With ReplayBuffer
[200] Accuracy: 84.00% – 0:00:00
[400] Accuracy: 86.25% – 0:00:00
[600] Accuracy: 87.50% – 0:00:00
[800] Accuracy: 88.25% – 0:00:00
[1,000] Accuracy: 89.20% – 0:00:00
[1,200] Accuracy: 89.75% – 0:00:00
Out[59]:
Accuracy: 89.68%
```

Without ReplayBuffer:

```python
from creme import linear_model
from creme import metrics
from creme import preprocessing
from creme import optim
from creme import datasets

model = (
     preprocessing.StandardScaler() |
    linear_model.LogisticRegression()
)

X_y = datasets.Phishing()

model_selection.progressive_val_score(
    X_y,
    model,
    metrics.Accuracy(),
    print_every=200,
    show_time=True
)
```

```
# Without ReplayBuffer
[200] Accuracy: 83.50% – 0:00:00
[400] Accuracy: 86.00% – 0:00:00
[600] Accuracy: 87.50% – 0:00:00
[800] Accuracy: 88.50% – 0:00:00
[1,000] Accuracy: 89.10% – 0:00:00
[1,200] Accuracy: 89.33% – 0:00:00
Out[45]:
Accuracy: 89.28%
```

#### datasets.TrumpApproval:

With ReplayBuffer:

```python
from creme import compose 
from creme import feature_extraction
from creme import stats
from creme import datasets
from creme import preprocessing
from creme import replay
from creme import linear_model
from creme import optim
from creme import metrics
from creme import model_selection

X_y = datasets.TrumpApproval()

model = preprocessing.StandardScaler()

model = (
    preprocessing.StandardScaler() | 
    replay.ReplayBufferRegressor(
        regressor = linear_model.LinearRegression(), 
        loss_function = optim.losses.Absolute(), 
        p = 0.1,
        size = 40, 
        seed = 42,
    )
)

model_selection.progressive_val_score(
    X_y, 
    model, 
    metrics.MAE(), 
    print_every=250
)
```

```
# MAE with ReplayBuffer
[250] MAE: 5.181268
[500] MAE: 3.045086
[750] MAE: 2.296498
[1,000] MAE: 1.849182
Out[11]:
MAE: 1.847611
```

Without ReplayBuffer

```python
from creme import compose 
from creme import feature_extraction
from creme import stats
from creme import datasets
from creme import preprocessing
from creme import linear_model
from creme import optim
from creme import metrics
from creme import model_selection

X_y = datasets.TrumpApproval()

model = preprocessing.StandardScaler()

model = (
    preprocessing.StandardScaler() | 
    linear_model.LinearRegression()
)

model_selection.progressive_val_score(
    X_y, 
    model, 
    metrics.MAE(), 
    print_every=250
)
```

```
# MAE without ReplayBuffer
[250] MAE: 5.393482
[500] MAE: 3.356939
[750] MAE: 2.577593
[1,000] MAE: 2.07941
Out[12]:
MAE: 2.077663
```

Raphaël :-)

### CODE REVIEW

File: creme/replay/base.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    To update the buffer, you must specify the loss of a given tuple (features, target). This new
+    observation will be stored if the loss is sufficiently high compared to the losses already
+    stored in the buffer.
+
+    Parameters:
+        size (int): Number of stored tuples (features, target).
+        p (float): Probability to update the model with an observation from the buffer when fitting
+            on a new observation. 0. <= p <= 1.
+        seed (int): Random state.
```
Comment (gbolmier) [inline creme/replay/base.py]: As we do in `optim.initializers`:
```
        random_state (int, ``numpy.random.RandomState`` instance or None): If int, ``random_state``
            is the seed used by the random number generator; if ``RandomState`` instance,
            ``random_state`` is the random number generator; if ``None``, the random number
            generator is the ``RandomState`` instance used by `numpy.random`.
```
See next comment below

File: creme/replay/base.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        self.p = p
+        self.buffer = {}
+        # Intiliaze the losses history with respect to the buffer size with losses equals to -1.
+        self.loss_history = {key: -1. for key in range(self.size)}
+        # Initialize the minimum loss observed to -1.
+        self.min_loss = -1.
+        # Initialize the key associated to the minimum loss observed.
+        self.key_min_loss = 0
+        self.seed = seed
+        self._rng = np.random.RandomState(seed)
```
Comment (gbolmier) [inline creme/replay/base.py]:
```
from sklearn import utils
self.random_state = utils.check_random_state(random_state)
```

**Comment by gbolmier** [comment]
Nice work @raphaelsty! Could you try to use `meta.PredClipper` to check that they work well together:

```python
from creme import datasets
from creme import linear_model
from creme import meta
from creme import metrics
from creme import model_selection
from creme import optim
from creme import preprocessing
from creme import replay

regressor = meta.PredClipper(
    regressor=linear_model.LinearRegression(),
    y_min=0,
    y_max=100
)

model = preprocessing.StandardScaler()
model |= replay.ReplayBufferRegressor(
    regressor=regressor,
    loss_function=optim.losses.Absolute(),
    p=0.1,
    size=40,
    seed=42
)

model_selection.progressive_val_score(
    datasets.TrumpApproval(), 
    model, 
    metrics.MAE(), 
    print_every=250
)
```
Results should be equal. Then, if you replace the interval with `y_min=36` and `y_max=45` we should get something different

File: creme/replay/base.py (hunk 2 above)
Comment (MaxHalford) [reply creme/replay/base.py]: Actually if possible I think it's preferable to use the standard library instead of depending on `sklearn` :).

**Comment by MaxHalford** [comment]
Lol @gbolmier, we should probably also check what happens with regards to the `graphviz` representation when we use nested wrapper models.

File: creme/replay/base.py (hunk 2 above)
Comment (gbolmier) [reply creme/replay/base.py]: Should we create our own `utils.check_random_state` and remove sklearn other dependencies?

Comment (MaxHalford) [reply creme/replay/base.py]: Using `self._rng = random.Random(seed)` works just fine. I prefer to `seed` to `random_state` namewise. We don't need any utility function!

File: creme/replay/replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    Parameters:
+        classifier (base.Classifier): Selected model.
+        loss_function (creme.optim.losses): Criterion to store observations in the buffer.
+        size (int): Number of stored tuples (features, target).
+        p (float): Probability to update the model with an observation from the buffer when fitting
+            on a new observation. 0. <= p <= 1.
+        seed (int): Random state.
+
+    """
+    def __init__(self, classifier, loss_function, size, p, seed=None):
```
Comment (raphaelsty) [inline creme/replay/replay_buffer.py]: The parameter in input of the class ReplayBufferClassifier for the model is "classifier". 

The parent class initialize it as "model" eg self.model = classifier.

Should I replace for ReplayBufferRegressor (eg, "regressor") and ReplayBufferClassifier (eg "classifier") by "model"?

**Comment by raphaelsty** [comment]
We are able to draw the model with the buffer. :-)

```python
from creme import compose 
from creme import feature_extraction
from creme import stats
from creme import datasets
from creme import preprocessing
from creme import replay
from creme import linear_model
from creme import optim
from creme import metrics
from creme import model_selection

X_y = datasets.TrumpApproval()

model = preprocessing.StandardScaler()

model = (
    preprocessing.StandardScaler() | 
    replay.ReplayBufferRegressor(
        regressor = linear_model.LinearRegression(), 
        loss_function = optim.losses.Absolute(), 
        p = 0.1,
        size = 40, 
        seed = 42,
    )
)

model_selection.progressive_val_score(
    X_y, 
    model, 
    metrics.MAE(), 
    print_every=250
)

import graphviz

model.draw()
```

image

**Comment by raphaelsty** [comment]
In order to clipper the model predictions, the buffer and the model must be nested inside the clipper. The meta.PredClipper and ReplayBufferRegressor classes overload the prediction methods, so there is a nesting order to be respected. @gbolmier :-)

```python
from creme import datasets
from creme import linear_model
from creme import meta
from creme import metrics
from creme import model_selection
from creme import optim
from creme import preprocessing
from creme import replay

regressor = meta.PredClipper(
    regressor = replay.ReplayBufferRegressor(
        regressor = linear_model.LinearRegression(),
        loss_function = optim.losses.Absolute(),
        p = 0.1,
        size = 40,
        seed = 42
    ),
    y_min = 13,
    y_max = 14
)

model = preprocessing.StandardScaler()
model |= regressor

model_selection.progressive_val_score(
    datasets.TrumpApproval(), 
    model, 
    metrics.MAE(), 
    print_every=250
)
```

**Comment by raphaelsty** [comment]
@gbolmier @MaxHalford @VaysseRobin @AdilZouitine 

The PR is ready for review.
